File: server/XSF/Module/NetPoint.cs
//////////////////////////////////////////////////////////////////////////
// 
// 文件：server/XSF/Module/NetPoint.cs
// 作者：Xoen Xie
// 时间：2023/07/01
// 描述：网络节点
// 说明：
//
//////////////////////////////////////////////////////////////////////////
#pragma warning disable CS8602, CS8618, CS8604

namespace XSF
{
    public class NetPoint : ITimerHandler, INetHandler
    {
        enum TimerID
        {
            HTCheck = 0,
            Max,
        }

        public ServerID SID { get { return m_SID; } }
        public uint ID { get { return m_nID; } }
        protected uint m_nID;
        protected ServerID m_SID;

        protected IConnection? m_Connection;
        protected uint m_nLastHTTime;

        public NPManager Owner { get; private set;}

        private TimersManager m_Timers;

        public bool Create(NPManager owner, IConnection connection)
        {
            Owner = owner;
            m_Connection = connection;
            m_nLastHTTime = XSFUtil.CurrentS;

            m_Timers = new TimersManager((int)TimerID.Max);

            m_Timers.StartTimer((byte)TimerID.HTCheck, this, XSFServer.Instance.Config.HeartbeatInterval, -1, "NetPoint.Create");

            return true;
        }

        public void SetID(uint nID)
        {
            m_nID = nID;
            m_SID = ServerID.GetSID(m_nID);
        }

        public void Release()
        {
            if(m_Connection != null)
            {
                m_Connection.Close();
            }

            m_Timers.CloseAllTimer();
        }


        public void OnTimer(byte nTimerID, bool bLastCall)
        {

        }

        public void UpdateHTTime()
        {
            m_nLastHTTime = XSFUtil.CurrentS;
        }

        public string RemoteIP
        {
            get
            {
                return m_Connection.RemoteIP;
            }
        }

        public void SendMessage(IMessage message)
        {
            if (m_Connection == null)
            {
                Serilog.Log.Error($"NetPoint.SendMessage connection not exist, msg id={message.ID}");
                return;
            }

            m_Connection.SendMessage(message);
        }

        public void SendData(ushort nMessageID, byte[] data)
        {
            if (m_Connection == null)
            {
                Serilog.Log.Error($"NetPoint.SendMessage connection not exist, msg id={nMessageID}");
                return;
            }

            m_Connection.SendData(data);
        }


        public void OnConnected(IConnection connection)
        {
            Serilog.Log.Error("NetPoint.OnConnected error call, name={0}", Owner.Name);
        }

        public void OnRecv(IConnection connection, IMessage message, ushort nMessageID, uint nRawID, byte[]? data)
        {
            message.Execute(this, nMessageID, nRawID, data);
        }

        public void OnError(IConnection connection, NetError nErrorCode)
        {
            Serilog.Log.Error("NetPoint.OnError, name={0}, id=[{1}|{2}-{3}-{4}]", Owner.Name, ID, SID.ID, XSFUtil.EP2CNName(SID.Type), SID.Index);
            Owner.OnNPError(this);
            Owner.Delete(this);
        }

        public void OnAccept(IConnection connection)
        {
            Serilog.Log.Error("NetPoint.OnAccept error call, name={0}", Owner.Name);
        }
    }
}

File: server-cpp/source/xsf/src/net_point/NetPoint.cpp
//////////////////////////////////////////////////////////////////////////
//
// 文件：base\src\net_point\XSFNetPoint.cpp
// 作者：Xoen Xie
// 时间：2023/08/14
// 描述：网络节点对象
// 说明：
//
//////////////////////////////////////////////////////////////////////////

#include "NetPoint.h"
#include "XSFLog.h"
#include "NetPointManager.h"
#include "XSF.h"
#include "XSFServer.h"

namespace xsf
{
    bool NetPoint::Create( INetPointManager * pOwner, IConnection * pConnection )
    {
        FUNCTION_ENTER(0);
        m_pOwner = pOwner;
        m_pConnection = pConnection;

        m_nLastHTTime = XSFCore::TickCount();

        m_Timers.StartTimer(TimerID_HTCheck, this, XSFServer::Instance()->GetConfig()->HeartbeatCheck, -1, "NetPoint::Create");

        FUNCTION_EXIT(0);

        return true;
    }

    void NetPoint::Close(void)
    {
        CONNECTION_CLOSE(m_pConnection);
        m_Timers.CloseAll();
    }

    void NetPoint::SendMessage( IMessage * pMessage )
    {
        if( m_pConnection == nullptr )
        {
            return;
        }

        m_pConnection->SendMessage(pMessage);
    }

    void NetPoint::SendData( const byte * pData, uint32 nDataLen )
    {
        if( m_pConnection != nullptr )
            m_pConnection->Send(pData, nDataLen);
    }

    void NetPoint::UpdateHTTime(void)
    {
        m_nLastHTTime = XSFCore::TickCount();
    }

    const char * NetPoint::GetRemoteIP(void)
    {
        return m_pConnection->GetRemoteIP();
    }




    // 接受到连接
    INetHandler * NetPoint::OnAccept(IConnection * pIncomingConn)
    {
        XSF_ERROR("XSFNetPoint::OnAccept illegal call, owner name:%s", m_pOwner->GetModuleName() );

        return nullptr;
    }

    // 已连接到远程主机
    void NetPoint::OnConnected( IConnection * pConection )
    {
        XSF_ERROR("XSFNetPoint::OnConnected illegal call, owner name:%s", m_pOwner->GetModuleName() );
    }

    // 连接出错或断开
    void NetPoint::OnError( IConnection * pConection, uint32 nErrorCode )
    {
        XSF_ERROR("NetPoint[%s], id:[%u|%u-%u-%u] error", m_pOwner->GetModuleName(), m_ID.ID, m_ID.S.server, m_ID.S.type, m_ID.S.index );
        
        m_Timers.CloseAll();

        m_pOwner->Delete(this);
    }

    // 收到数据
    void NetPoint::OnRecv(IConnection * pConection, DataResult * pResult)
    {
        IMessage * pMessage = XSFCore::GetMessage(pResult->nMessageID);
        if(pMessage != nullptr)
        {
            pMessage->Import(pResult->pPBData, pResult->nPBLength);

            pMessage->Execute(this, pMessage, pResult);
        }
        else
        {
            XSF_WARN("NetPoint::OnRecv pMessage is null, id=%u", pResult->nMessageID);
        }

        UpdateHTTime();
    }


    TIMER_FUNCTION_START(NetPoint)
        (TIMER_CALL)&NetPoint::OnTimerHTCheck,
    TIMER_FUNCTION_END


    void NetPoint::OnTimerHTCheck(bool bLastCall)
    {
        uint32 nNow = XSFCore::TickCount();
        if( nNow > m_nLastHTTime + XSFServer::Instance()->GetConfig()->HeartbeatTimeout )
        {
            XSF_ERROR("NetPoint::OnTimerHTCheck NetPoint time out, time diff(nNow:%u-m_nLastHTTime:%u):%u, owner name:%s", nNow, m_nLastHTTime, nNow-m_nLastHTTime, m_pOwner->GetModuleName());
            Close();

            m_pOwner->Delete(this);
        }
    }

}